src/test/java/org/jenkinsci/plugins/mesos/MesosComputerLauncherTest.java

```diff
@@ -0,0 +1,103 @@
+        Timer.Context  timerContext   = Mockito.mock(Timer.Context.class);
+
+        Mockito.when(mesosCloud.getRole()).thenReturn("root");
+        Mockito.when(mesos.isSchedulerRunning()).thenReturn(true);
+        Mockito.when(Mesos.getInstance(Matchers.any())).thenReturn(mesos);
+        Mockito.when(Metrics.metricRegistry()).thenReturn(metricRegistry);
+        Mockito.when(metricRegistry.timer(Matchers.anyString()).time())
+                .thenReturn(timerContext);
+    }
+
+
+    /**
+     * The launch method can enter an infinite loop when a computer is
+     * coming online, fails, and never connects. This test exercises those
+     * conditions and ensures the method exits as expected.
+     */
+    @Test
+    public void testLaunchExitsOnFailure() throws InterruptedException {
+        final long     startTime      = System.currentTimeMillis();
+        MesosComputer  mesosComputer  = Mockito.mock(MesosComputer.class);
+        MesosSlave     mesosSlave     = Mockito.mock(MesosSlave.class, Mockito.RETURNS_DEEP_STUBS);
+        MesosSlaveInfo mesosSlaveInfo = Mockito.mock(MesosSlaveInfo.class);
+        TaskListener   taskListener   = Mockito.mock(TaskListener.class);
+        PrintStream    printStream    = Mockito.mock(PrintStream.class);
+
+        Mockito.when(taskListener.getLogger()).thenReturn(printStream);
+        Mockito.when(mesosComputer.getNode()).thenReturn(mesosSlave);
+        Mockito.when(mesosComputer.isOffline()).thenReturn(true);
+        Mockito.when(mesosComputer.isConnecting()).thenReturn(true);
+        Mockito.when(mesosComputer.isOnline()).thenReturn(true);
+        Mockito.when(mesosSlave.getCpus()).thenReturn(1.0);
+        Mockito.when(mesosSlave.getMem()).thenReturn(32);
+        Mockito.when(mesosSlave.getDiskNeeded()).thenReturn(100.0);
+        Mockito.when(mesosSlave.getSlaveInfo()).thenReturn(mesosSlaveInfo);
+        Mockito.when(mesosSlave.getProvisioningContext().stop())
+                .thenReturn(System.currentTimeMillis() - startTime);
+
+        // place to store the SlaveResult object passed on
+        // `startJenkinsSlave`
+        final Mesos.SlaveResult[] result = new Mesos.SlaveResult[1];
+
+        Mockito.doAnswer(invocation -> {
+            final Mesos.SlaveResult slaveResult = (Mesos.SlaveResult) (invocation.getArguments())[1];
+            slaveResult.running(null);
+            result[0] = slaveResult;
+            return null;
+        })
+                .when(mesos)
+                .startJenkinsSlave(Matchers.any(Mesos.SlaveRequest.class),
+                        Matchers.any(Mesos.SlaveResult.class));
+
+        MesosComputerLauncher launcher = new MesosComputerLauncher(mesosCloud, "mycloud");
+        Thread                t        = new Thread(() -> launcher.launch(mesosComputer, taskListener));
+        t.setName("mesos-launcher-thread");
+        t.start();
+        assert (t.isAlive());
+        // need to wait for some things to progress
+        Thread.sleep(250L);
```

Comment:
It would be good to avoid sleeps if possible.  There are utilities like `awaitility` or perhaps some restructuring of the product code to support an unthreaded mode would be appropriate.